### PR TITLE
Dev/ft/ring 28537 data placement v1

### PR DESCRIPTION
**Pull request by maximelubin**
Landing smarter, more flexible hyperdrive selection scheme

### CODE REVIEW

**Comment by maximelubin** [comment]
To all reviewers: I recommend going over it commit by commit, in-order except for the last 2 commits which you may want to read first since they are about the documentation and design.

File: docs/Design.rst [hunk 1]
```diff
@@ -231,7 +231,102 @@ emit an error and early close the streamed data.
 Data placement
 --------------
 
-TODO
+Since the have several hyperdrives to store data on, the question of how to select them on a put. Ideally
+the seelction algorthm should perform akin to a wear-levelling SSD algorithm, essentially smoothing
+keys over hyerpdrives depending on their current capacity: available space, status, etc.
```
Comment (benoit-a) [inline docs/Design.rst]: a few typos (seelction algorthm hyerpdrives

File: docs/Design.rst [hunk 2]
```diff
@@ -231,7 +231,102 @@ emit an error and early close the streamed data.
...
+- staticWeight: positive number, indirectly describing the attractiveness of a hyperdrive. Static because
+  this weight does not take into account liveness inputs.
+- dynamicWeight: corrected weight with liveness info.
+- affinity: 'soft'=multiple fragments or 'hard'=single. Leaves default to 'hard' while higher levels default on 'soft'.
+- ftype: 'data', 'coding' or 'both' (default) specifying which kind of fragment is accepted.
+
+Containers follow roughly the same pattern, except they have a list of children (components), no staticWeight,
+and their dynamicWeight is the sum of the weights of their children.
+
+Example: (XX) are hyperdirve weights, [XX] are aggregated weights
```
Comment (benoit-a) [inline docs/Design.rst]: typo

File: lib/config.js [hunk 3]
```diff
@@ -35,55 +224,52 @@ function validatePolicySection(policy) {
...
+    if (! (policy.cluster.components instanceof Array)) {
         return {
             configIsValid: false,
             configError: new InvalidConfigError(
-                'policy.locations',
-                policy.locations,
-                'Expected an array of string'),
+                'policy.cluster.components',
+                policy.cluster.components,
+                'A clsuter expects at least 1 described component'),
```
Comment (benoit-a) [inline lib/config.js]: typo

File: lib/config.js [hunk 4]
```diff
@@ -214,6 +397,7 @@ function validate(opts) {
                 'Expected a positive number'),
         };
     }
+    config.requestTimeoutMs = opts.requestTimeoutMs;
```
Comment (benoit-a) [inline lib/config.js:400]: Side note: timeout == 0 is ok ?

Comment (maximelubin) [reply lib/config.js:400]: Yes. 0 means no timeout

File: docs/Design.rst [hunk 5]
```diff
@@ -231,7 +231,102 @@ emit an error and early close the streamed data.
 Data placement
 --------------
 
-TODO
+Since the have several hyperdrives to store data on, the question of how to select them on a put. Ideally
```
Comment (jgagnaire) [inline docs/Design.rst:234]: "Since there are several hyperdrives" ?

**Review by jgagnaire — APPROVED** [review]

**Pull request merged**